### PR TITLE
Fork merge: Diagram-Text tracelinks, Improvements to the MME detection capabilities

### PR DESCRIPTION
**Pull request by menof36go**
These changes are the result of my thesis _Entity Recognition in Software Documentation Using Trace Links to Informal Diagrams_

The last merge with main was commit d653a1381556dad1cd87b2c02b748e8f75b7818d. Some integration tests will probably fail because of the minor global changes. Generally, a favorable change is observed for the metrics. However, a small worsening in a particular metric might fail the tests.

Please let me know about any changes I should perform or oversights (e.g. missing documentation of classes I created). Before the final merge (assuming this PR is approved), I will update to the latest commit on ArDoCo main. Due to the large amount of changed files, it is likely that I missed some stuff.

Best regards,
Fabian

### CODE REVIEW

**Comment by dfuchss** [comment]
@menof36go could you apply the formatter / spotless :) `mvn spotless:apply`

Maybe this will resolve diffs like that:
image

**Comment by menof36go** [comment]
> @menof36go could you apply the formatter / spotless :) `mvn spotless:apply`
> 
> Maybe this will resolve diffs like that: image

I was previously unaware of the maven spotless plugin. I applied the command that is applied by github actions, namely 

> mvn -B spotless:apply --file pom.xml

However, no issues could be found. I also tried with check instead, and I tried running both commands without additional arguments as well. I attached the log file for the aforementioned command. To me it looks identical to the github actions process, but it does not find any issues. [spotless_apply_log.txt](https://github.com/ArDoCo/Core/files/13060676/spotless_apply_log.txt)

Do you have any idea what might cause the discrepancy? I am still using 0.26.0 as a baseline, however the mvn spotless plugin version is identical with 2.4.0. The only difference that is obvious to me is that it did not download the dependencies (it did however do this the first time I executed the goal, so I think this is not the issue)

**Comment by dfuchss** [comment]
I just recognized that you've worked on a main branch :)

Just remove try to remove https://github.com/ArDoCo/Core/blob/84ace0267ab53b642809c72bea6b3b092b395902/pom.xml#L467 locally or rename the branch. I'll discuss with the other maintainers whether we can remove https://github.com/ArDoCo/Core/blob/84ace0267ab53b642809c72bea6b3b092b395902/pom.xml#L467 in future development ..

**Review by dfuchss — COMMENTED** [review]
I've reviewed 100/340 files. Good work so far :) I have some comments to some classes.

File: stages/diagram-recognition/src/main/kotlin/edu/kit/kastel/mcse/ardoco/lissa/diagramrecognition/informants/DockerInformant.kt
Comment (dfuchss) [inline stages/diagram-recognition/src/main/kotlin/edu/kit/kastel/mcse/ardoco/lissa/diagramrecognition/informants/DockerInformant.kt:21]: Please add a comment what this is

File: stages/diagram-recognition/src/main/kotlin/edu/kit/kastel/mcse/ardoco/lissa/diagramrecognition/informants/OcrInformant.kt
Comment (dfuchss) [inline stages/diagram-recognition/src/main/kotlin/edu/kit/kastel/mcse/ardoco/lissa/diagramrecognition/informants/OcrInformant.kt:37]: Please revert

File: stages/diagram-recognition/src/main/kotlin/edu/kit/kastel/mcse/ardoco/lissa/diagramrecognition/informants/OcrInformant.kt
Comment (dfuchss) [inline stages/diagram-recognition/src/main/kotlin/edu/kit/kastel/mcse/ardoco/lissa/diagramrecognition/informants/OcrInformant.kt:105]: Why always recreate?

File: stages/inconsistency-detection/src/main/java/edu/kit/kastel/mcse/ardoco/core/inconsistency/informants/RecommendedInstanceProbabilityFilter.java
Comment (dfuchss) [inline stages/inconsistency-detection/src/main/java/edu/kit/kastel/mcse/ardoco/core/inconsistency/informants/RecommendedInstanceProbabilityFilter.java:108]: Probably revert this change

File: stages/pom.xml
Comment (dfuchss) [inline stages/pom.xml:15]: Move this to complete and other profiles

**Review by dfuchss — COMMENTED** [review]
I reviewed the next 90 classes

File: framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/api/Disambiguation.java
Comment (menof36go) [reply framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/api/Disambiguation.java:32]: As in `java private final Set<String> meanings`? I think it might be better to be specific. Set makes no guarantee about iteration order by itself and doesn't necessarily give information about mutability either (most people and ides just assume immutability). Thus, I think the interfaces Set and Map shouldn't be used, if we want to comply with the determinism architecture test. If I were to use Set here, it would additionally require annotating the class `@Deterministic` (at least that's my understanding of the annotation). This seems like a long-term maintenance headache to me, since it relies on manual review and technically needs to be checked manually on every change.

I do however recognize the need for immutable/unmodifiable (deterministic) collections, which the interfaces are often used to represent implicitly (which is poor design by Java itself imo, sadly the explicitly unmodifiable collections are package-private in java.util). In such cases, I propose using the classes [UnmodifiableLinkedHashSet](https://github.com/ArDoCo/Core/blob/bd8235ac3c5f9e07531773197cf4b60785e28d3f/framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/common/collection/UnmodifiableLinkedHashSet.java) and [UnmodifiableLinkedHashMap](https://github.com/ArDoCo/Core/blob/bd8235ac3c5f9e07531773197cf4b60785e28d3f/framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/common/collection/UnmodifiableLinkedHashMap.java) instead. These classes offer deterministic iteration order and explicitly specify mutability. They also implement Set and Map and thus do not break any existing code where Set or Map is used. The alternative solution is using sorted eclipse collections everywhere, which would require implementing comparables and inducing performance penalties.

| Type/Feature  | Mutability | Determinism | Compatible with built-in types |
| ------------- | ------------- | ------------- | ------------- |
| Set/Map | No | No | Yes |
| Eclipse Collections | Yes | No | Only mutable |
| Eclipse Collections Sorted | Yes | Yes | Only mutable |
| (Unmodifiable-)LinkedHashSet/Map | Yes  | Yes | Yes |

File: framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/api/diagramrecognition/Classification.java
Comment (menof36go) [reply framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/api/diagramrecognition/Classification.java]: I honestly have no clue. Probably something vestigial that i accidentally merged from some older version? Ill remove it

File: framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/api/diagramrecognition/DiagramElement.java
Comment (menof36go) [reply framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/api/diagramrecognition/DiagramElement.java:129]: Why not? Equals and hashcode solely depend on immutable values

File: framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/api/Disambiguation.java
Comment (menof36go) [reply framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/api/Disambiguation.java]: The signature has been adapted to use (Unmodifiable-)LinkedHashMaps. I can change it if you want sorted parameters and return values, but if the concern was mainly about determinism, this may be sufficient

File: framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/api/diagramrecognition/DiagramElement.java
Comment (menof36go) [reply framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/api/diagramrecognition/DiagramElement.java:93]: Fixed for all occurrences of LazyInitializer

File: framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/api/textextraction/NounMapping.java
Comment (menof36go) [reply framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/api/textextraction/NounMapping.java:69]: I dont think so. In my understanding conceptually a phrase should only exist once, since it is a specific sequence of words from the text. I think I changed this for consistency reasons, since words are also stored in ImmutableSortedSets. I also faintly remember this being the reason for some rather large slowdown when using the ArDoCo Util Comparators functions which I discovered when I did some profiling to mitigate performance issues. Do you have anything in particular in mind that I should check regarding whether or not it broke?

File: framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/api/models/ModelConnector.java
Comment (menof36go) [reply framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/api/models/ModelConnector.java:11]: This has the advantage that changes to the codebase invalidate persistent datarepository caches (since java automatically changes serialversionuid if major changes occur), which reduces unexpected behaviour due to caching. It has the disadvantage of requiring some memory and some performance for serialization. However, both is rather negligible compared to the effort required for serializing things such as the words, sentences and phrases. It is a tradeoff in that sense. If required, I will change it and mark the required fields transient, it is up to you

File: framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/common/util/Comparators.java
Comment (menof36go) [reply framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/common/util/Comparators.java:76]: I am not sure what you mean. This is just an overload of a function that is supposed to check whether two collections contain identical elements. If two sorted iterables contain identical elements, the statement "Both collections contain identical elements in any order" still holds. This overload is merely for the special case where the provided collections are already sorted, because it speeds up the check. Or am I missing something? I don't see a scenario where this is a problem

I did however spot a bug just now. Are the functions supposed to ignore cardinality? Because as of now, the default implementation will return true for two collections such as:

Coll A: A-A-B-C
Coll B: A-B-B-C

Im not sure that this is intended by the original author of the functions, but if it is I need to adapt the overload slightly

File: framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/pipeline/agent/Claimant.java
Comment (menof36go) [reply framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/pipeline/agent/Claimant.java:9]: See [above](https://github.com/ArDoCo/Core/pull/286#discussion_r1378715552)

File: stages/connection-generator/src/main/java/edu/kit/kastel/mcse/ardoco/core/connectiongenerator/informants/ProjectNameInformant.java
Comment (menof36go) [reply stages/connection-generator/src/main/java/edu/kit/kastel/mcse/ardoco/core/connectiongenerator/informants/ProjectNameInformant.java:55]: projectName is derived from ProjectPipelineData, which is populated from the ArDoCoRunner, which provides the name in its constructor. However, these classes are sparsely documented. For the tests, the name is either in some cases hardcoded as a constant `"teastore"` in the test class or derived from the enum name. Since the enum name is always UPPERCASE, it is immediately converted to lowercase in most classes before being passed to the runner. 

This is very easy to miss. If you do not provide the name in lowercase, the project name informant will not filter the project name correctly, leading to a large worsening in the metrics. I can attest to this, since at one point i missed converting it to lowercase before feeding it to the runner and spend the next 60+ minutes trying to figure out why my metrics are terrible.

However, i think it is more sensible to adjust the ProjectNameInformant so that it ignores case when searching for similar words. I will do that and remove line 55, if that's fine

File: framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/api/diagramrecognition/Classification.java
Comment (menof36go) [reply framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/api/diagramrecognition/Classification.java]: Nevermind. I am completely confused. This is currently already part of [Classification](https://github.com/ArDoCo/Core/blob/main/framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/api/diagramrecognition/Classification.java). I have no clue why this even shows as an addition, since it is part of current main branch already. It seems to be something you added about 3 months ago in commit 8630c2bf2ec1f207ce8100332696d82cdf710bb8

File: stages/diagram-connection-generator/src/main/java/edu/kit/kastel/mcse/ardoco/erid/api/diagramconnectiongenerator/DiagramConnectionState.java
Comment (menof36go) [reply stages/diagram-connection-generator/src/main/java/edu/kit/kastel/mcse/ardoco/erid/api/diagramconnectiongenerator/DiagramConnectionState.java:31]: No, it doesnt. I moved it into the implementation, which extends AbstractConfigurable anyways

File: stages/diagram-recognition/pom.xml
Comment (menof36go) [reply stages/diagram-recognition/pom.xml:137]: When i try to compile the module without this using `mvn clean compile -pl io.github.ardoco.core:diagram-recognition -am`, the compile fails because maven tries to compile the java classes before compiling the kotlin classes.

The kotlin compile is executed after the java default-compile. I think kotlin [needs to be compiled first](https://discuss.kotlinlang.org/t/compiling-mixed-java-and-kotlin-files-on-the-command-line/1553).

I think this is a configuration issue that no one stumbled upon so far, because there are no other mixed java / kotlin modules (I think?). I have no clue how to resolve this, since my knowledge about kotlin is minimal. Probably somehow the order needs to be changed in parent pom?

File: framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/api/diagramrecognition/Classification.java
Comment (dfuchss) [reply framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/api/diagramrecognition/Classification.java]: Ok . Maybe the diff was wrong :)

File: stages/diagram-recognition/src/main/kotlin/edu/kit/kastel/mcse/ardoco/lissa/diagramrecognition/informants/OcrInformant.kt
Comment (menof36go) [reply stages/diagram-recognition/src/main/kotlin/edu/kit/kastel/mcse/ardoco/lissa/diagramrecognition/informants/OcrInformant.kt:37]: Id like to also set this with some sort of environment variable. Maybe this can even be inferred from context? This is true in all cases except when REMOTE_DOCKER_URI is set, right? If that's the case, Ill just check that env var to determine the appropriate value

**Comment by dfuchss** [comment]
Currently, the CI gives `Runtime java.io.NotSerializableException: edu.kit.kastel.mcse.ardoco.core.textproviderjson.textobject.DependencyImpl`

File: tests/tests-base/src/main/resources/benchmark/bigbluebutton/diagrams_2015/README.md
Comment (dfuchss) [inline tests/tests-base/src/main/resources/benchmark/bigbluebutton/diagrams_2015/README.md:1]: Could we move the Benchmark stuff into a PR to Benchmark? The Benchmark is linked via git-subtree :)

**Comment by menof36go** [comment]
I fixed some sonar cloud issues, fixed the DtoObjectConversion (it would break due to the brackets in the text not being escaped prior to trying to parse the dependency tree). Also fixed some other issues and improved performance. 

Also finally removed the AmbiguationInformant. Turns out my "TODO" was right all along and it can just be removed.

The sonar cloud security hotspot is a non-issue, since we are dealing with small inputs.

Im not sure how to fix the remaining sonar cloud "bugs", since I am not a regex expert in general

**Comment by menof36go** [comment]
CI fail is caused by the changes to 2015 teammates.txt in [this commit](https://github.com/menof36go/Benchmark/commit/e1799f861687f12dfe62d5057bf810089fe6228f). My understanding of the benchmark text files is that every line is supposed to be a sentence, which is why I added the period that was missing in the first place

**Comment by dfuchss** [comment]
@menof36go after holiday / in the new year, could you resolve the conflicts ? :) The progress of this PR is good; soonish we hopefully can merge it :)

**Comment by dfuchss** [comment]
@menof36go if you find some time please have a look at the PR :) We plan to adjust the architecture of ArDoCo soon and before that we want to merge the PR :)

File: tests/pom.xml
Comment (dfuchss) [inline tests/pom.xml:18]: Revert :)

File: tests/tests-inconsistency/pom.xml
Comment (dfuchss) [inline tests/tests-inconsistency/pom.xml:91]: Why do you need the test-jar?

File: tests/tests-resources/pom.xml
Comment (dfuchss) [inline tests/tests-resources/pom.xml:10]: What does this module do?

File: framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/api/diagramrecognition/Box.java
Comment (menof36go) [reply framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/api/diagramrecognition/Box.java:129]: Box now extends DiagramElement which in turn extends entity. In my understanding entity names are unique and thus the function has now kind of become obsolete (Since we also set the name to be the UUID, they are definitely in this case). The function should probably be removed entirely and getName can be used instead

File: framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/api/models/arcotl/code/CodeItem.java
Comment (menof36go) [reply framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/api/models/arcotl/code/CodeItem.java:46]: I removed this during profiling, because the function is only called in the CodeItemRepository init on the values of the repository. It is never used in a context where a CodeItem is unregistered. The Code SAM Tlr stuff was really slow, mainly because the CodeItemRepository kept being reinitialized, which is very costly when you consider that it is sorted and its size. So I made sure that it would properly be initialized only once and that unnecessary add calls are removed
Long story short: Its intentional, not breaking and increases performance

File: framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/api/textextraction/NounMappingChangeListener.java
Comment (menof36go) [reply framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/api/textextraction/NounMappingChangeListener.java:6]: Yes, since we do not necessarily know when a state is serialized, we want to preserve relationships as best as we can. Imagine a scenario like the Inconsistency Detection, where we want to run the same stage X times using the same "base run"/"pre run". To achieve proper separation between runs using a deep copy is required, which necessitates serializability of all things belonging to a state, including relationships between state elements.

File: stages/code-traceability/src/main/java/edu/kit/kastel/mcse/ardoco/core/codetraceability/SamCodeTraceabilityLinkRecovery.java
Comment (menof36go) [reply stages/code-traceability/src/main/java/edu/kit/kastel/mcse/ardoco/core/codetraceability/SamCodeTraceabilityLinkRecovery.java:29]: Because of "global" structures. Static classes, singletons, etc are very bad for testing (especially in parallel). In the past there was a missing distinction between "actually global and stateless" and "global on a per-run basis". For example, WordSimUtils used to be a "static" class, however it is not stateless. In truth, it is a "global on a per-run basis" kind of class, that is shared between stages, but can be configured. 

So we have to find a solution to make such "global but not stateless" classes accessible to the stages without infringing on the separation between runs. The solution I came up with was attaching these classes to a thing that is already on a per-run basis -> the datarepository. Instead of referencing a static configurable class (bad), we can now make sure each run has its own instance in the datarepository, which it can configure as it pleases. However, since we can no longer just reference such a class statically, its instance has to be accessible to all constituents of the run. For example, the TextStateImpl used to statically refer to the sim utils, which it cannot do anymore. By having access to the datarepository, it can get the sim utils instance from it instead.

Ofcourse, in the particular case that you highlighted, the code traceability state currently doesn't make use of this. For an example of how this may be used, have a look at one of these stages

grafik

I hope I explained it reasonably well, I suck at explaining

File: tests/tests-resources/pom.xml
Comment (menof36go) [reply tests/tests-resources/pom.xml:10]: It is responsible for handling the diagram gold standard stuff. E.g. it parses the gold standard. It also contains the "StageTest" that the test classes of my stages implement (Runs repetitions of the stage to test determinism, provides caching functionality of test results for comparison, etc)

File: stages/code-traceability/src/main/java/edu/kit/kastel/mcse/ardoco/core/codetraceability/SamCodeTraceabilityLinkRecovery.java
Comment (dfuchss) [reply stages/code-traceability/src/main/java/edu/kit/kastel/mcse/ardoco/core/codetraceability/SamCodeTraceabilityLinkRecovery.java:29]: mmh .. I get it. I'll think about it :)

File: tests/tests-inconsistency/pom.xml
Comment (menof36go) [reply tests/tests-inconsistency/pom.xml:91]: I wanted to subclass the classes used for evaluation that this model provides. However, since the classes are in src/test, they are not part of the build and cannot be used as a dependency without telling maven to create a test-jar. Classes in the test subdirectory are supposed to test classes in the main subdirectory. The tlr and id test modules violate these principles. In retrospect, fixing this problem by moving the main code into the main subdirectory seems like a more elegant and sensible solution. After that, the snippet above can be removed

File: framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/api/models/arcotl/code/CodeItem.java
Comment (dfuchss) [reply framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/api/models/arcotl/code/CodeItem.java:46]: Did you also tested the recreation of code models or only the loading?

Comment (dfuchss) [reply framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/api/models/arcotl/code/CodeItem.java:46]: Seems to be safe :)

File: framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/pipeline/Pipeline.java
Comment (menof36go) [reply framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/pipeline/Pipeline.java:107]: Sure why not. I did not take into account the possibility of running a pipeline instance multiple times

File: framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/pipeline/ExecutionStage.java
Comment (menof36go) [reply framework/common/src/main/java/edu/kit/kastel/mcse/ardoco/core/pipeline/ExecutionStage.java:10]: Yes, I think the purpose of this class was to get rid of having a static "get" function in each stage where you create a new stage and initialize it with the given config (because such a function cannot be part of the abstract class). In hindsight, it makes more sense to just adjust the original abstract class (add additional configs as a constructor parameter)  instead and remove this class. Or use archrules to enforce the creation of a static get method in non-abstract classes extending abstract execution stage (for the sake of consistency)

**Review by dfuchss — APPROVED** [review]
I think we have resolved the remaining major issues :)

**Pull request merged**
